colossalai/fx/passes/autoparallel/runtime_preparation_pass.py

```diff
@@ -0,0 +1,130 @@
+
+            for name, param in target_module.named_parameters():
+                target_sharding_spec = node.best_strategy.get_sharding_spec_by_name(name)
+                # apply the sharding spec of parameters
+                if target_sharding_spec.dim_partition_dict != {}:
+                    origin_sharding_spec = ShardingSpec(device_mesh, param.shape, {})
+                    setattr(param, 'sharding_spec', origin_sharding_spec)
+                    param_sharded = torch.nn.Parameter(
+                        shape_consistency_manager.apply_for_autoparallel_runtime(param.data, param.sharding_spec,
+                                                                                 target_sharding_spec).detach().clone())
+                else:
+                    param_sharded = param
+                setattr(target_module, name, param_sharded)
+                comm_actions = node.best_strategy.communication_actions
+                for operation_data, comm_action in comm_actions.items():
+                    comm_spec_to_use = comm_action.comm_spec
+                    # register hook to the parameters
+                    if operation_data.type == OperationDataType.PARAM and operation_data.name == name and comm_action.comm_type == CommType.HOOK:
+
+                        def wrapper(param, comm_spec):
+
+                            def hook_fn(grad):
+                                _all_reduce(grad, comm_spec)
+
+                            param.register_hook(hook_fn)
+
+                        wrapper(param_sharded, comm_spec_to_use)
+
+            sharded_buffer_dict = {}
+            # apply the sharding spec of buffers
+            for name, buffer in target_module.named_buffers():
+                origin_sharding_spec = ShardingSpec(device_mesh, buffer.shape, {})
+                setattr(buffer, 'sharding_spec', origin_sharding_spec)
+                target_sharding_spec = node.best_strategy.get_sharding_spec_by_name(name)
+                buffer_sharded = shape_consistency_manager.apply(buffer, target_sharding_spec)
+                sharded_buffer_dict[name] = buffer_sharded
+
+            for name, buffer_sharded in sharded_buffer_dict.items():
+                setattr(target_module, name, buffer_sharded.detach().clone())
+
+    return gm
+
+
+def implicit_comm_action_apply_pass(gm: torch.fx.GraphModule):
+    """
+    replace the origin kernel into kernel with implicit communication inside.
+    """
+    pass
+
+
+def runtime_preparation_pass(gm: torch.fx.GraphModule, solution: List[int], device_mesh: DeviceMesh):
+    gm, sharding_spec_convert_dict, origin_node_sharding_spec_dict, comm_actions_dict = _solution_annotatation_pass(
+        gm, solution)
+    # TODO: the pass below should be uncommented after the implementation of implicit_comm_action_apply_pass completed.
+    # gm = implicit_comm_action_apply_pass(gm)
+    gm = _module_params_sharding_pass(gm, device_mesh)
```

Comment:
The function names seem a bit weird to me, it is kind of saying that there are passes within passes as `runtime_preparation_pass` calls `_solution_annotatation_pass` and `_module_params_sharding_pass`. Would it better if we treat `_module_params_sharding_pass` and `_solution_annotatation_pass` as the helper functions instead of passes.

Reply:
I see.